File: frontend/app/src/main/java/com/asflum/cashewregister/google/gmail/GmailExpenseSyncManager.kt
package com.asflum.cashewregister.google.gmail

import android.content.Context
import com.asflum.cashewregister.BackendProcessor
import com.asflum.cashewregister.FileDownloader

object GmailExpenseSyncManager {

    suspend fun syncAndDownloadExpenses(context: Context, idToken: String): Result<String> {
        val messagesResult = GmailService.readMessages(idToken)
        if (messagesResult.isFailure) return messagesResult

        val messages = messagesResult.getOrNull() ?: return Result.failure(Exception("TODO"))

        val processResult = BackendProcessor.processExpenses(messages)
        if (processResult.isFailure) return processResult

        val messagesProcessed = processResult.getOrNull() ?: return Result.failure(Exception("TODO"))

        val downloadResult = FileDownloader.downloadToDevice(context, messagesProcessed)
        if (downloadResult.isFailure) return downloadResult

        return downloadResult
    }
}

File: frontend/lib/services/gmail_expense_sync_manager.dart
import 'package:expense_tracker/utils/result.dart';
import 'package:expense_tracker/services/gmail_service.dart';
import 'package:expense_tracker/services/file_downloader.dart';

class GmailExpenseSyncManager {
  static Future<Result<String>> syncAndDownloadExpenses(
    String sessionToken,
  ) async {
    // Obtener los datos del backend (ya procesados con categorías)
    final messagesResult = await GmailService.readMessages(sessionToken);

    if (messagesResult.isFailure) {
      return Result.failure(messagesResult.exceptionOrNull()!);
    }

    final csvBytes = messagesResult.getOrNull()!;

    // Descargar el archivo CSV
    final downloadResult = await FileDownloader().downloadToDevice(csvBytes);

    return downloadResult;
  }
}
